File: legacy-express-backend/controllers/authController.js
const User = require("../models/User");
const jwt = require("jsonwebtoken");

/**
 * Generate JWT token for an authenticated user
 */
const generateToken = (id) => {
    return jwt.sign({ id }, process.env.JWT_SECRET, {
        expiresIn: "30d",
    });
};

/**
 * @desc    Register a new user
 * @route   POST /api/auth/signup
 * @access  Public
 */
const registerUser = async (req, res) => {
    try {
        const { name, email, phone, password } = req.body;

        // 1. Validate inputs
        if (!name || !email || !phone || !password) {
            return res.status(400).json({
                success: false,
                message: "Please provide all required fields: name, email, phone, password",
            });
        }

        // Basic email formatting validation helper
        const emailRegex = /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/;
        if (!emailRegex.test(email)) {
            return res.status(400).json({
                success: false,
                message: "Please enter a valid email address",
            });
        }

        if (password.length < 6) {
            return res.status(400).json({
                success: false,
                message: "Password must be at least 6 characters long",
            });
        }

        // 2. Check if user already exists
        const userExists = await User.findOne({ email });
        if (userExists) {
            return res.status(400).json({
                success: false,
                message: "User already exists with this email address",
            });
        }

        // 3. Create user
        const user = await User.create({
            name,
            email,
            phone,
            password,
        });

        if (user) {
            return res.status(201).json({
                success: true,
                message: "User registered successfully",
                data: {
                    _id: user._id,
                    name: user.name,
                    email: user.email,
                    phone: user.phone,
                    token: generateToken(user._id),
                },
            });
        } else {
            return res.status(400).json({
                success: false,
                message: "Invalid user data received",
            });
        }
    } catch (error) {
        return res.status(500).json({
            success: false,
            message: "Server error occurred during signup",
            error: error.message,
        });
    }
};

/**
 * @desc    Authenticate user & get token
 * @route   POST /api/auth/login
 * @access  Public
 */
const loginUser = async (req, res) => {
    try {
        const { email, password } = req.body;

        // 1. Validate inputs
        if (!email || !password) {
            return res.status(400).json({
                success: false,
                message: "Please provide both email and password",
            });
        }

        // 2. Check for user
        const user = await User.findOne({ email });
        if (!user) {
            return res.status(401).json({
                success: false,
                message: "Invalid email or password",
            });
        }

        // 3. Check if password matches
        const isMatch = await user.comparePassword(password);
        if (!isMatch) {
            return res.status(401).json({
                success: false,
                message: "Invalid email or password",
            });
        }

        // 4. Return user and token
        return res.json({
            success: true,
            message: "Logged in successfully",
            data: {
                _id: user._id,
                name: user.name,
                email: user.email,
                phone: user.phone,
                token: generateToken(user._id),
            },
        });
    } catch (error) {
        return res.status(500).json({
            success: false,
            message: "Server error occurred during login",
            error: error.message,
        });
    }
};

module.exports = {
    registerUser,
    loginUser,
};

File: src/main/java/com/foodexpress/controller/AuthController.java
package com.foodexpress.controller;

import com.foodexpress.model.User;
import com.foodexpress.repository.UserRepository;
import com.foodexpress.security.JwtTokenProvider;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.web.bind.annotation.*;
import java.util.HashMap;
import java.util.Map;
import java.util.regex.Pattern;

@RestController
@RequestMapping("/api/auth")
public class AuthController {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider tokenProvider;

    private static final Pattern EMAIL_PATTERN = Pattern.compile("^\\w+([\\.-]?\\w+)*@\\w+([\\.-]?\\w+)*(\\.\\w{2,3})+$");

    public AuthController(UserRepository userRepository,
                          PasswordEncoder passwordEncoder,
                          JwtTokenProvider tokenProvider) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
        this.tokenProvider = tokenProvider;
    }

    @PostMapping("/signup")
    public ResponseEntity<Map<String, Object>> signup(@RequestBody SignupRequest request) {
        Map<String, Object> response = new HashMap<>();

        // 1. Validate inputs
        if (request.getName() == null || request.getName().trim().isEmpty() ||
            request.getEmail() == null || request.getEmail().trim().isEmpty() ||
            request.getPhone() == null || request.getPhone().trim().isEmpty() ||
            request.getPassword() == null || request.getPassword().trim().isEmpty()) {
            response.put("success", false);
            response.put("message", "Please provide all required fields: name, email, phone, password");
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
        }

        // Email validation
        if (!EMAIL_PATTERN.matcher(request.getEmail()).matches()) {
            response.put("success", false);
            response.put("message", "Please enter a valid email address");
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
        }

        // Domain validation: enforce @gmail.com except for dev@example.com
        String emailToCheck = request.getEmail().trim().toLowerCase();
        if (!emailToCheck.equals("dev@example.com") && !emailToCheck.endsWith("@gmail.com")) {
            response.put("success", false);
            response.put("message", "Only Gmail addresses (@gmail.com) are allowed to register.");
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
        }

        // Password length check
        if (request.getPassword().length() < 6) {
            response.put("success", false);
            response.put("message", "Password must be at least 6 characters long");
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
        }

        // 2. Check if user already exists
        if (userRepository.findByEmail(request.getEmail().toLowerCase()).isPresent()) {
            response.put("success", false);
            response.put("message", "User already exists with this email address");
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
        }

        // 3. Create and save user
        User user = new User();
        user.setName(request.getName());
        user.setEmail(request.getEmail().toLowerCase());
        user.setPhone(request.getPhone());
        user.setPassword(passwordEncoder.encode(request.getPassword()));

        User savedUser = userRepository.save(user);

        // Generate token and build response
        String token = tokenProvider.generateToken(savedUser.getId());

        Map<String, Object> userData = new HashMap<>();
        userData.put("_id", savedUser.getId());
        userData.put("name", savedUser.getName());
        userData.put("email", savedUser.getEmail());
        userData.put("phone", savedUser.getPhone());
        userData.put("role", savedUser.getRole());
        userData.put("token", token);

        response.put("success", true);
        response.put("message", "User registered successfully");
        response.put("data", userData);

        return ResponseEntity.status(HttpStatus.CREATED).body(response);
    }

    @PostMapping("/login")
    public ResponseEntity<Map<String, Object>> login(@RequestBody LoginRequest request) {
        Map<String, Object> response = new HashMap<>();

        // 1. Validate inputs
        if (request.getEmail() == null || request.getEmail().trim().isEmpty() ||
            request.getPassword() == null || request.getPassword().trim().isEmpty()) {
            response.put("success", false);
            response.put("message", "Please provide both email and password");
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
        }

        // 2. Check for user
        User user = userRepository.findByEmail(request.getEmail().toLowerCase()).orElse(null);
        if (user == null) {
            response.put("success", false);
            response.put("message", "Invalid email or password");
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(response);
        }

        // 3. Check password
        if (!passwordEncoder.matches(request.getPassword(), user.getPassword())) {
            response.put("success", false);
            response.put("message", "Invalid email or password");
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(response);
        }

        // Generate token and build response
        String token = tokenProvider.generateToken(user.getId());

        Map<String, Object> userData = new HashMap<>();
        userData.put("_id", user.getId());
        userData.put("name", user.getName());
        userData.put("email", user.getEmail());
        userData.put("phone", user.getPhone());
        userData.put("role", user.getRole());
        userData.put("token", token);

        response.put("success", true);
        response.put("message", "Logged in successfully");
        response.put("data", userData);

        return ResponseEntity.ok(response);
    }

    // Helper requests DTO classes
    public static class SignupRequest {
        private String name;
        private String email;
        private String phone;
        private String password;

        public String getName() { return name; }
        public void setName(String name) { this.name = name; }
        public String getEmail() { return email; }
        public void setEmail(String email) { this.email = email; }
        public String getPhone() { return phone; }
        public void setPhone(String phone) { this.phone = phone; }
        public String getPassword() { return password; }
        public void setPassword(String password) { this.password = password; }
    }

    public static class LoginRequest {
        private String email;
        private String password;

        public String getEmail() { return email; }
        public void setEmail(String email) { this.email = email; }
        public String getPassword() { return password; }
        public void setPassword(String password) { this.password = password; }
    }
}
